### DISCUSSION TITLE
LibSQL/Turso support as database driver

### DISCUSSION THREAD

**Discussion by BrandonDR** [laravel/framework#49340] [Ideas]
LibSQL is an open-source fork of SQLite. It powers the growing [Turso](https://docs.turso.tech/) platform, which provides distributed database support. Whereas, SQLite does NOT handle asynchronous operations well making SQLite unsuitable for production environments at moderate scale. 

Currently, there appears to be no first-party LibSQL PHP driver. However, given it's similarities to the currently supported SQLite driver, I wonder if we can add support for LibSQL in Laravel.

LibSQL allows for the SQLite dialect of SQL over HTTP, as well as `.sqlite` file databases. Making it great for server-less deployments and for simplifying CI/CD deployment for Laravel unit/integration tests. Simply copy a prepared database file and you have a repeatable environment for testing. LibSQL appears to solve the production downsides of SQLite with its replication and asynchronous support.

This all means, we can have fast integration tests that can closely replicate the production environment with same SQL dialect without running a heavy test MySQL server, for example. Which would have its own maintenance, state and deployment complexity.

LibSQL seems very promising and developer friendly.

#### Turso / LibSQL resources
- https://github.com/tursodatabase/libsql
- https://turso.tech/libsql
- https://turso.tech/

I think the implementation would roughly be:
1. Extract the existing SQL from the Laravel/Eloquent SQLite driver
2. Adapt SQL to run over LibSQL, based on configuration this may be over the HTTP transport to the [LibSQL server](https://github.com/tursodatabase/libsql/tree/main/libsql-server)

Thank you for your consideration. I would be interested in helping efforts towards this support.

**Comment by defenestrator** [comment]
I'm taking a crack at building a package/library for this, starting today. For the simple reason that I love Turso and Laravel, I would like to use them together. I'm quite glad I searched up this issue. Is anyone aware of ongoing efforts toward this end? @BrandonDR I was thinking, same as you, of just building an HTTP client on the Turso API, rather than a PHP extension for LibSQL, but possibly both, and offer a config option for it down the road. I don't see any reason why the http client would be terribly complex to implement, I may reach out to you with a request for contributions after I get a POC going.

**Reply by BrandonDR** [reply]
I don't believe anyone has started this yet. Yeah, I will try to help out once you have the POC. I look forward to see it

**Comment by defenestrator** [comment]
https://github.com/darkterminal/turso-http - this exists, I'll be be trying it out, it is listed as an Experimental SDK in the Turso docs, it appeared somewhat recently and looks pretty good.

**Comment by zwinglio** [comment]
i think that could be a greate adition!

**Comment by peterfox** [comment]
As someone who made a custom DB driver for Eloquent (cockroachDB driver) I will say, making the driver is fairly easy but the hardest bit is grabbing the tests from the Laravel Repo.

Eloquent is built on top of the PHP PDO extension. There might be some difficulty with that as I assume you'd want to implement it 100% in PHP instead of trying to write a PDO extension.

Overally it wouldn't be too hard as all the Grammar is just SQLite.

**Reply by darkterminal** [reply]
I don't know so much about Laravel since version 4 :rofl:, but here is I am writing [Native LibSQL Extension](https://github.com/laravel/framework/discussions/49340?sort=old#discussioncomment-9480404) that support Turso Database features.

**Comment by nick-potts** [comment]
I started messing about with it here - https://github.com/nick-potts/Laravel-LibSQL

I think the best implementation method is extending the PDO and StatementPDO and using the HTTP client there. That way you keep the stock Laravel implementation of sqlite.

**Comment by nick-potts** [comment]
I've got it pretty close now, there's something funky that I haven't been able to figure out with transactions rolling back, but the base concept is working.

Edit: transactions working now, the only last thing is binary blob storage isn't working

**Reply by BrandonDR** [reply]
@nick-potts Looking good, I'm going to try it out. Great job

**Comment by defenestrator** [comment]
I started consuming nick-potts' libsql for Laravel implementation yesterday, was looking into the BLOB bug and updating the README, but as of today it has been archived. 

Check out `https://github.com/richan-fongdasen/turso-laravel` - it looks, off the rip, pretty good to me!

**Reply by nick-potts** [reply]
Yep, decided to deprecate mine in favour of his. I hacked mine together to prove it worked, but he's made far more progress in the meanwhile.

**Comment by darkterminal** [comment]
Hello Punk!

I build bunch of Turso/LibSQL Client for fun experiment and I love it! I am also build:

- [TursoHTTP](https://github.com/darkterminal/turso-http) (that mentioned by @defenestrator) it's connect Remote Turso Database via HTTP
- [TursoSyncd](https://github.com/darkterminal/turso-syncd) it's stupid worker that sync local database to remote database via HTTP and it will make you poor for a long usage :sweat_smile:
- [LibSQL Client PHP](https://github.com/darkterminal/libsql-client-php) Turso + PHP SDK for Native LibSQL via FFI it's only support **Local Connection**, **Remote Connection** over HTTP, and **Remote Replica**. Also you can check this at [Turso Client SDK](https://docs.turso.tech/sdk/php/quickstart)
- [LibSQL PHP Ext](https://github.com/darkterminal/libsql-php-ext) LibSQL PHP Extension via FFI - Core Dependency for LibSQL Client PHP. This extension behave like SQLite driver but it's LibSQL!

---

## The Final

I've done building a Native LibSQL Extension/Driver/Whatever and I am try to package this driver into Laravel (but I need some help here :smile:). Since this is not an official PHP extension, I need to make it installable in the project.

## What I've Done?

This weekend May, 19 2024, I am done and re-think about what I am talking before. I just drunk stupid person but still; I am handsome. The [LibSQL PHP Ext](https://github.com/darkterminal/libsql-php-ext) can be use stand-alone, this is the core dependency of [LibSQL Client PHP](https://github.com/darkterminal/libsql-client-php).

## Requirements

- Linux or Darwin OS
- C/C++ Compiler
- jq
- Rust Installed
- PHP Installed
- FFI Extension is Enabled (_Why? I read the C heder definition from wrapper_)

## 🚨 DON'T WORRY ABOUT REQUIREMENTS 🚨

Everything will be prepared when you install this extension. But, if you encountered an error; create an issue.

So with [LibSQL PHP Ext](https://github.com/darkterminal/libsql-php-ext) features we can create custom **LibSQL Driver** for Laravel (but it's not support remote), remote connection still handle via HTTP SDK from [LibSQL Client PHP](https://darkterminal/libsql-client-php).

So let me know, who want to join the party :tada: :rocket: 

> NOTE This party only for Linux & Mac/Darwin. But Windows can join when using WSL :smile: 

## Another Options

I am also create native extension [LibSQL Extension](https://github.com/darkterminal/libsql-extension), so you can download and install them and configured in `php.ini` but still, it's not based on PDO driver.

## Recent Updates

Make sure you follow the latest update from [My X Thread](https://x.com/panggilmeiam/status/1793206655024496968)!

<details>
<summary>22 Mei 2024 - LibSQL Driver Laravel - Initial</summary>
I created <a href="https://github.com/darkterminal/libsql-driver-laravel" target="_blank">LibSQL Driver Laravel</a> still work in progress. It's will used <code>Illuminate\Support\Facades\DB</code> to interact with LibSQL native driver.
</details>

<details>
<summary>23 Mei 2024 - LibSQL Driver Laravel - LibSQL Local Connection</summary>
And now <a href="https://github.com/darkterminal/libsql-driver-laravel" target="_blank">LibSQL Driver Laravel</a> can perform database operations using traditional <code>Illuminate\Support\Facades\DB</code> for LibSQL Local Connection:

- Migration and Seeding
- CRUD
- Transaction

</details>

<details>
<summary>23 Mei 2024 - LibSQL Driver Laravel - LibSQL Remote Replica Connection</summary>
And now <a href="https://github.com/darkterminal/libsql-driver-laravel" target="_blank">LibSQL Driver Laravel</a> can perform database <code>sync</code> operations using traditional <code>Illuminate\Support\Facades\DB</code> for LibSQL Remote Replica Connection:

- Migration and Seeding
- CRUD
- Transaction
- Sync

Here the screenshot of fresh Laravel app database migration and data:

1. **Local `database.sqlite` tables**

![Local `database.sqlite` tables](https://github.com/laravel/framework/assets/32319439/2836c371-7e99-47a6-9e62-f496a236dad9)

2. **Remote database tables at Turso Dashboard**

![Remote database tables at Turso Dashboard](https://github.com/laravel/framework/assets/32319439/3869bac5-8351-4f52-accb-172a00e09fdf)

3. **Users Data from Seeder Database at Turso Dashboard**

![Users Data from Seeder Database at Turso Dashboard](https://github.com/laravel/framework/assets/32319439/0976c9d6-1dea-4fe0-a5ae-cf697f278ae2)

5. **Users Data From in Local `database.sqlite` File**

![Users Data From in Local `database.sqlite` File](https://github.com/laravel/framework/assets/32319439/14aef558-beb0-4ecb-bcea-a9ace529ed62)

</details>

LibSQL Drive Laravel - Out Now!!!

![LibSQL Drive Laravel](https://github.com/laravel/framework/assets/32319439/4513a837-7b68-431f-a3e6-e0a144355b37)

> Tweet: https://x.com/panggilmeiam/status/1793857451969896480
